graphql_compiler/deserialization.py

```diff
@@ -0,0 +1,253 @@
+            GraphQLFloat: float
+            GraphQLDecimal: decimal.Decimal
+            GraphQLInt: int
+            GraphQLString: str
+            GraphQLBoolean: bool
+            GraphQLID: str
+            GraphQLList: list of the inner type
+
+    Raises:
+        GraphQLInvalidArgumentError: if the argument value was not of the expected type.
+    """
+    stripped_type = strip_non_null_from_type(expected_type)
+    if isinstance(stripped_type, GraphQLList):
+        if not isinstance(value, list):
+            raise GraphQLInvalidArgumentError(
+                f"Attempted to convert argument {name} to a GraphQLList, but {name} had a non-list "
+                f"value {value}."
+            )
+
+        inner_stripped_type = strip_non_null_from_type(stripped_type.of_type)
+
+        return [
+            deserialize_scalar_argument(name, inner_stripped_type, element) for element in value
+        ]
+    else:
+        return deserialize_scalar_argument(name, stripped_type, value)
+
+
+def deserialize_multiple_arguments(
+    arguments: Mapping[str, Any], expected_types: Mapping[str, QueryArgumentGraphQLType],
+) -> Dict[str, Any]:
+    """Deserialize serialized GraphQL arguments.
+
+    Passing arguments via jsonrpc, or via the GUI of standard GraphQL editors is tricky because
+    json does not support certain types like Date, Datetime, Decimal, and also confuses floats
+    for integers if there are no decimals. This function takes in values converts them to
+    standard python representations.
+
+    Below are examples of accepted json encodings of all the types:
+        GraphQLDate: "2018-02-01"
+        GraphQLDateTime: "2018-02-01T05:11:54Z"
+        GraphQLFloat: 4.3, "5.0", 5
+        GraphQLDecimal: "5.00000000000000000000000000001"
+        GraphQLInt: 4, "3803330000000000000000000000000000000000000000000"
+        GraphQLString: "Hello"
+        GraphQLBoolean: True, 1, "1", "True", "true"
+        GraphQLID: "13d72846-1777-6c3a-5743-5d9ced3032ed"
+        GraphQLList(GraphQLInt): [1, 2, 3]
+
+    Args:
+        arguments: mapping of argument names to serialized argument values.
+        expected_types: mapping of argument names to the expected GraphQL types. All
+                        GraphQLNonNull wrappers are stripped.
+
+    Returns:
+        a mapping of argument names to deserialized argument values. The type of the deserialized
+        argument value depends on the argument's GraphQL type:
+            GraphQLDate: datetime.date
+            GraphQLDateTime: datetime.datetime with tzinfo=pytz.utc
+            GraphQLFloat: float
+            GraphQLDecimal: decimal.Decimal
+            GraphQLInt: int
+            GraphQLString: str
+            GraphQLBoolean: bool
+            GraphQLID: str
+            GraphQLList: list of the inner type
+
+    Raises:
+        GraphQLInvalidArgumentError: if any of the argument values was not of the expected type.
+    """
+    return {
```

Comment:
Can you re-add the `ensure_arguments_are_provided` functionality here?

Reply:
I opted to remove because I didn't want to import from `query_formatting`. This function was only used in 1 instance in `test_end_to_end` so I just moved the call to `ensure_arguments_are_provided` to that test. Alternatively, I could make a function in `query_formatting/common.py` called `validate_and_deserialize_multiple_arguments` or something similar that calls `ensure_arguments_are_provided` and then this function. Since the function was only used once, I didn't think it was worthwhile to create such a function, but happy to put one in if you think it would be beneficial.

Reply:
I don't think we should add a `validate_and_deserialize_multiple_arguments` since it doesn't make sense to have two functions are almost the same except for the one difference that one performs more validation. (When would someone ever call `deserialize_multiple_arguments`). Also, I think that some sort of validation is already implied whenever a function is parsing/deserializing something. 

However, I do think that it's extremely important to have `ensure_arguments_are_provided_here`.  Suppose that the user inputs additional or misspelled arguments. Then this function will raise a `KeyError` instead of a `GraphQLInvalidArgumentError`. While it is safe to regard `GraphQLInvalidArgumentError`  as being due to invalid user input, a `KeyError` will most likely generate some sort of alarm if this code is used in production. This alarm will probably waste developer productivity since this is most likely a user error and nothing can be done. 

The missing arguments case is probably worse because there will still be some sort of error/alarm raised but it won't be as obvious that it was raised because of faulty user input.

I agree though that we should probably not import this function from `query_formatting/common.py`. I think that it makes sense to move this function elsewhere. (Maybe `global_utils.py`). I know that this function is part of the public API of that module but I am not sure if we've released a new compiler version since that function was introduced. Do you mind checking? If it is part of the public API then we can still move it elsewhere and import it to `query_formatting/common.py` to avoid breaking API.